File: ts/sdk/src/jitProxyClient.ts
import {
	BN,
	DriftClient,
	getSwiftUserAccountPublicKey,
	isVariant,
	MakerInfo,
	MarketType,
	PostOnlyParams,
	QUOTE_SPOT_MARKET_INDEX,
	ReferrerInfo,
	TxParams,
	UserAccount,
} from '@drift-labs/sdk';
import { IDL, JitProxy } from './types/jit_proxy';
import {
	PublicKey,
	SYSVAR_INSTRUCTIONS_PUBKEY,
	TransactionInstruction,
	TransactionMessage,
	VersionedTransaction,
} from '@solana/web3.js';
import { Program } from '@coral-xyz/anchor';
import { TxSigAndSlot } from '@drift-labs/sdk';
import { SignedSwiftOrderParams } from '@drift-labs/sdk/lib/node/swift/types';

export type JitIxParams = {
	takerKey: PublicKey;
	takerStatsKey: PublicKey;
	taker: UserAccount;
	takerOrderId: number;
	maxPosition: BN;
	minPosition: BN;
	bid: BN;
	ask: BN;
	postOnly: PostOnlyParams | null;
	priceType?: PriceType;
	referrerInfo?: ReferrerInfo;
	subAccountId?: number;
};

export type JitSwiftIxParams = JitIxParams & {
	authorityToUse: PublicKey;
	signedSwiftOrderParams: SignedSwiftOrderParams;
	uuid: Uint8Array;
	marketIndex: number;
};

export class PriceType {
	static readonly LIMIT = { limit: {} };
	static readonly ORACLE = { oracle: {} };
}

export type OrderConstraint = {
	maxPosition: BN;
	minPosition: BN;
	marketIndex: number;
	marketType: MarketType;
};

export class JitProxyClient {
	private driftClient: DriftClient;
	private program: Program<JitProxy>;

	constructor({
		driftClient,
		programId,
	}: {
		driftClient: DriftClient;
		programId: PublicKey;
	}) {
		this.driftClient = driftClient;
		this.program = new Program(IDL, programId, driftClient.provider);
	}

	public async jit(
		params: JitIxParams,
		txParams?: TxParams
	): Promise<TxSigAndSlot> {
		const ix = await this.getJitIx(params);
		const tx = await this.driftClient.buildTransaction([ix], txParams);
		return await this.driftClient.sendTransaction(tx);
	}

	public async jitSwift(
		params: JitSwiftIxParams,
		txParams?: TxParams,
		precedingIxs?: TransactionInstruction[]
	): Promise<TxSigAndSlot> {
		const swiftTakerIxs = await this.driftClient.getPlaceSwiftTakerPerpOrderIxs(
			params.signedSwiftOrderParams,
			params.marketIndex,
			{
				taker: params.takerKey,
				takerStats: params.takerStatsKey,
				takerUserAccount: params.taker,
			},
			params.authorityToUse,
			precedingIxs
		);

		const ix = await this.getJitSwiftIx(params);
		const tx = await this.driftClient.buildTransaction(
			[...swiftTakerIxs, ix],
			txParams
		);
		let resp;
		try {
			const message = new TransactionMessage({
				payerKey: this.driftClient.wallet.payer.publicKey,
				recentBlockhash: (
					await this.driftClient.connection.getLatestBlockhash()
				).blockhash,
				instructions: [...swiftTakerIxs, ix],
			}).compileToV0Message([this.driftClient.lookupTableAccount]);

			const tx = new VersionedTransaction(message);
			resp = await this.driftClient.connection.simulateTransaction(tx, {
				sigVerify: false,
				replaceRecentBlockhash: true,
				commitment: 'processed',
			});
			console.log(resp);
		} catch (e) {
			console.error(e);
		}

		return await this.driftClient.sendTransaction(tx);
	}

	public async getJitIx({
		takerKey,
		takerStatsKey,
		taker,
		takerOrderId,
		maxPosition,
		minPosition,
		bid,
		ask,
		postOnly = null,
		priceType = PriceType.LIMIT,
		referrerInfo,
		subAccountId,
	}: JitIxParams): Promise<TransactionInstruction> {
		subAccountId =
			subAccountId !== undefined
				? subAccountId
				: this.driftClient.activeSubAccountId;
		const order = taker.orders.find((order) => order.orderId === takerOrderId);
		const remainingAccounts = this.driftClient.getRemainingAccounts({
			userAccounts: [taker, this.driftClient.getUserAccount(subAccountId)],
			writableSpotMarketIndexes: isVariant(order.marketType, 'spot')
				? [order.marketIndex, QUOTE_SPOT_MARKET_INDEX]
				: [],
			writablePerpMarketIndexes: isVariant(order.marketType, 'perp')
				? [order.marketIndex]
				: [],
		});

		if (referrerInfo) {
			remainingAccounts.push({
				pubkey: referrerInfo.referrer,
				isWritable: true,
				isSigner: false,
			});
			remainingAccounts.push({
				pubkey: referrerInfo.referrerStats,
				isWritable: true,
				isSigner: false,
			});
		}

		if (isVariant(order.marketType, 'spot')) {
			remainingAccounts.push({
				pubkey: this.driftClient.getSpotMarketAccount(order.marketIndex).vault,
				isWritable: false,
				isSigner: false,
			});
			remainingAccounts.push({
				pubkey: this.driftClient.getQuoteSpotMarketAccount().vault,
				isWritable: false,
				isSigner: false,
			});
		}

		const jitParams = {
			takerOrderId,
			maxPosition,
			minPosition,
			bid,
			ask,
			postOnly,
			priceType,
		};

		return this.program.methods
			.jit(jitParams)
			.accounts({
				taker: takerKey,
				takerStats: takerStatsKey,
				state: await this.driftClient.getStatePublicKey(),
				user: await this.driftClient.getUserAccountPublicKey(subAccountId),
				userStats: this.driftClient.getUserStatsAccountPublicKey(),
				driftProgram: this.driftClient.program.programId,
			})
			.remainingAccounts(remainingAccounts)
			.instruction();
	}

	public async getJitSwiftIx({
		takerKey,
		takerStatsKey,
		taker,
		maxPosition,
		minPosition,
		bid,
		ask,
		postOnly = null,
		priceType = PriceType.LIMIT,
		referrerInfo,
		subAccountId,
		uuid,
		marketIndex,
	}: JitSwiftIxParams): Promise<TransactionInstruction> {
		subAccountId =
			subAccountId !== undefined
				? subAccountId
				: this.driftClient.activeSubAccountId;
		const remainingAccounts = this.driftClient.getRemainingAccounts({
			userAccounts: [taker, this.driftClient.getUserAccount(subAccountId)],
			writableSpotMarketIndexes: [],
			writablePerpMarketIndexes: [marketIndex],
		});

		if (referrerInfo) {
			remainingAccounts.push({
				pubkey: referrerInfo.referrer,
				isWritable: true,
				isSigner: false,
			});
			remainingAccounts.push({
				pubkey: referrerInfo.referrerStats,
				isWritable: true,
				isSigner: false,
			});
		}

		const jitSwiftParams = {
			swiftOrderUuid: Array.from(uuid),
			maxPosition,
			minPosition,
			bid,
			ask,
			postOnly,
			priceType,
		};

		return this.program.methods
			.jitSwift(jitSwiftParams)
			.accounts({
				taker: takerKey,
				takerStats: takerStatsKey,
				takerSwiftUserOrders: getSwiftUserAccountPublicKey(
					this.driftClient.program.programId,
					takerKey
				),
				authority: this.driftClient.wallet.payer.publicKey,
				state: await this.driftClient.getStatePublicKey(),
				user: await this.driftClient.getUserAccountPublicKey(subAccountId),
				userStats: this.driftClient.getUserStatsAccountPublicKey(),
				driftProgram: this.driftClient.program.programId,
				ixSysvar: SYSVAR_INSTRUCTIONS_PUBKEY,
			})
			.remainingAccounts(remainingAccounts)
			.instruction();
	}

	public async getCheckOrderConstraintIx({
		subAccountId,
		orderConstraints,
	}: {
		subAccountId: number;
		orderConstraints: OrderConstraint[];
	}): Promise<TransactionInstruction> {
		subAccountId =
			subAccountId !== undefined
				? subAccountId
				: this.driftClient.activeSubAccountId;

		const readablePerpMarketIndex = [];
		const readableSpotMarketIndexes = [];
		for (const orderConstraint of orderConstraints) {
			if (isVariant(orderConstraint.marketType, 'perp')) {
				readablePerpMarketIndex.push(orderConstraint.marketIndex);
			} else {
				readableSpotMarketIndexes.push(orderConstraint.marketIndex);
			}
		}

		const remainingAccounts = this.driftClient.getRemainingAccounts({
			userAccounts: [this.driftClient.getUserAccount(subAccountId)],
			readableSpotMarketIndexes,
			readablePerpMarketIndex,
		});

		return this.program.methods
			.checkOrderConstraints(orderConstraints)
			.accounts({
				user: await this.driftClient.getUserAccountPublicKey(subAccountId),
			})
			.remainingAccounts(remainingAccounts)
			.instruction();
	}

	public async arbPerp(
		params: {
			makerInfos: MakerInfo[];
			marketIndex: number;
		},
		txParams?: TxParams
	): Promise<TxSigAndSlot> {
		const ix = await this.getArbPerpIx(params);
		const tx = await this.driftClient.buildTransaction([ix], txParams);
		return await this.driftClient.sendTransaction(tx);
	}

	public async getArbPerpIx({
		makerInfos,
		marketIndex,
		referrerInfo,
	}: {
		makerInfos: MakerInfo[];
		marketIndex: number;
		referrerInfo?: ReferrerInfo;
	}): Promise<TransactionInstruction> {
		const userAccounts = [this.driftClient.getUserAccount()];
		for (const makerInfo of makerInfos) {
			userAccounts.push(makerInfo.makerUserAccount);
		}

		const remainingAccounts = this.driftClient.getRemainingAccounts({
			userAccounts,
			writablePerpMarketIndexes: [marketIndex],
		});

		for (const makerInfo of makerInfos) {
			remainingAccounts.push({
				pubkey: makerInfo.maker,
				isWritable: true,
				isSigner: false,
			});
			remainingAccounts.push({
				pubkey: makerInfo.makerStats,
				isWritable: true,
				isSigner: false,
			});
		}

		if (referrerInfo) {
			const referrerIsMaker =
				makerInfos.find((maker) =>
					maker.maker.equals(referrerInfo.referrer)
				) !== undefined;
			if (!referrerIsMaker) {
				remainingAccounts.push({
					pubkey: referrerInfo.referrer,
					isWritable: true,
					isSigner: false,
				});
				remainingAccounts.push({
					pubkey: referrerInfo.referrerStats,
					isWritable: true,
					isSigner: false,
				});
			}
		}

		return this.program.methods
			.arbPerp(marketIndex)
			.accounts({
				state: await this.driftClient.getStatePublicKey(),
				user: await this.driftClient.getUserAccountPublicKey(),
				userStats: this.driftClient.getUserStatsAccountPublicKey(),
				driftProgram: this.driftClient.program.programId,
			})
			.remainingAccounts(remainingAccounts)
			.instruction();
	}
}

File: python/sdk/jit_proxy/jit_proxy_client.py
from dataclasses import dataclass
from typing import Optional, cast

from borsh_construct.enum import _rust_enum
from sumtypes import constructor # type: ignore

from solders.pubkey import Pubkey # type: ignore

from anchorpy import Context, Program

from solana.transaction import AccountMeta

from driftpy.types import (
    UserAccount,
    PostOnlyParams,
    ReferrerInfo,
    MarketType,
    MakerInfo,
    is_variant,
)
from driftpy.drift_client import DriftClient
from driftpy.constants.numeric_constants import QUOTE_SPOT_MARKET_INDEX


@_rust_enum
class PriceType:
    Limit = constructor()
    Oracle = constructor()


@dataclass
class JitIxParams:
    taker_key: Pubkey
    taker_stats_key: Pubkey
    taker: UserAccount
    taker_order_id: int
    max_position: int
    min_position: int
    bid: int
    ask: int
    price_type: Optional[PriceType]
    referrer_info: Optional[ReferrerInfo]
    sub_account_id: Optional[int]
    post_only: PostOnlyParams = PostOnlyParams.MustPostOnly()


@dataclass
class ArbIxParams:
    maker_infos: list[MakerInfo]
    market_index: int


@dataclass
class OrderConstraint:
    max_position: int
    min_position: int
    market_index: int
    market_type: MarketType


class JitProxyClient:
    def __init__(self, drift_client: DriftClient, program_id: Pubkey):
        self.program_id = program_id
        self.drift_client = drift_client
        self.program = None

    async def init(self):
        self.program = await Program.at(
            self.program_id, self.drift_client.program.provider
        )

    async def jit(self, params: JitIxParams):
        if self.program is None:
            await self.init()

        sub_account_id = self.drift_client.get_sub_account_id_for_ix(
            params.sub_account_id # type: ignore
        )

        order = next(
            (
                order
                for order in params.taker.orders
                if order.order_id == params.taker_order_id
            ),
            None,
        )
        remaining_accounts = self.drift_client.get_remaining_accounts(
            user_accounts=[
                params.taker,
                self.drift_client.get_user_account(sub_account_id),
            ],
            writable_spot_market_indexes=[order.market_index, QUOTE_SPOT_MARKET_INDEX] # type: ignore
            if is_variant(order.market_type, "Spot") # type: ignore
            else [],
            writable_perp_market_indexes=[order.market_index] # type: ignore
            if is_variant(order.market_type, "Perp") # type: ignore
            else [],
        )

        if params.referrer_info is not None:
            remaining_accounts.append(
                AccountMeta(
                    pubkey=params.referrer_info.referrer,
                    is_writable=True,
                    is_signer=False,
                )
            )
            remaining_accounts.append(
                AccountMeta(
                    pubkey=params.referrer_info.referrer_stats,
                    is_writable=True,
                    is_signer=False,
                )
            )

        if is_variant(order.market_type, "Spot"): # type: ignore
            remaining_accounts.append(
                AccountMeta(
                    pubkey=self.drift_client.get_spot_market_account( # type: ignore
                        order.market_index # type: ignore
                    ).vault,
                    is_writable=False,
                    is_signer=False,
                )
            )
            remaining_accounts.append(
                AccountMeta(
                    pubkey=self.drift_client.get_quote_spot_market_account().vault, # type: ignore
                    is_writable=False,
                    is_signer=False,
                )
            )

        jit_params = self.program.type["JitParams"]( # type: ignore
            taker_order_id=params.taker_order_id,
            max_position=cast(int, params.max_position),
            min_position=cast(int, params.min_position),
            bid=cast(int, params.bid),
            ask=cast(int, params.ask),
            price_type=self.get_price_type(params.price_type), # type: ignore
            post_only=self.get_post_only(params.post_only),
        )

        ix = self.program.instruction["jit"]( # type: ignore
            jit_params,
            ctx=Context(
                accounts={
                    "state": self.drift_client.get_state_public_key(),
                    "user": self.drift_client.get_user_account_public_key(
                        sub_account_id
                    ),
                    "user_stats": self.drift_client.get_user_stats_public_key(),
                    "taker": params.taker_key,
                    "taker_stats": params.taker_stats_key,
                    "authority": self.drift_client.wallet.public_key,
                    "drift_program": self.drift_client.program_id,
                },
                signers={self.drift_client.wallet}, # type: ignore
                remaining_accounts=remaining_accounts,
            ),
        )

        tx_sig_and_slot = await self.drift_client.send_ixs(ix)

        return tx_sig_and_slot.tx_sig

    def get_price_type(self, price_type: PriceType):
        if is_variant(price_type, "Oracle"):
            return self.program.type["PriceType"].Oracle() # type: ignore
        elif is_variant(price_type, "Limit"):
            return self.program.type["PriceType"].Limit() # type: ignore
        else: 
            raise ValueError(f"Unknown price type: {str(price_type)}")
        
    def get_post_only(self, post_only: PostOnlyParams):
        if is_variant(post_only, "MustPostOnly"):
            return self.program.type["PostOnlyParam"].MustPostOnly() # type: ignore
        elif is_variant(post_only, "TryPostOnly"):
            return self.program.type["PostOnlyParam"].TryPostOnly() # type: ignore
        elif is_variant(post_only, "Slide"):
            return self.program.type["PostOnlyParam"].Slide() # type: ignore
